lib/dubhash.ts

```diff
@@ -0,0 +1,194 @@
+
+const hashAlgorithm = 'md5';
+
+const regex = /=/g;
+
+enum LineEndingFormat {
+  LINUX,
+  WINDOWS,
+  UNKNOWN,
+}
+
+export const getDubHashSignature = (
+  filePath: string,
+  fileContents: Buffer,
+  getAltHash = false,
+): SignatureResult | null => {
+  // Unless getAltHash is true, we only need a single hash, AKA the "half double hash".
+  if (!getAltHash) {
+    return {
+      path: filePath,
+      hashes_ffm: [
+        {
+          data: getSingleDigest(fileContents),
+          format: 1,
+        },
+      ],
+    };
+  }
+
+  const [digest1, digest2] = getDubHashDigests(
+    fileContents,
+    isBinary(fileContents),
+  );
+  if (!digest1) throw Error(`Failed to generate hash of ${filePath}`);
+  if (digest1 && !digest2) {
+    return {
+      path: filePath,
+      hashes_ffm: [
+        {
+          data: digest1,
+          format: 1,
+        },
+      ],
+    };
+  }
+  return {
+    path: filePath,
+    hashes_ffm: [
+      {
+        data: digest1,
+        format: 1,
+      },
+      {
+        data: digest2 as string, // digest2 is guaranteed not to be null, due to earlier return
+        format: 1,
+      },
+    ],
+  };
+};
+
+const getDubHashDigests = (
+  fileContents: Buffer,
+  isBinary: boolean,
+): Array<string | null> => {
+  // If the file is binary, you only produce one hash
+  // If the file does not have line endings, you treat it like a binary file
+  // If the LF character is not found, the line endings are UNKNOWN, and we produce only one hash.
+  const lineEndingFormat: LineEndingFormat = detectLineEndingFormat(
+    fileContents,
+  );
+  if (
+    isBinary ||
+    !hasLineEndings(fileContents) ||
+    lineEndingFormat === LineEndingFormat.UNKNOWN
+  ) {
+    return [getSingleDigest(fileContents), null];
+  }
+  // NB: You MUST calculate the digest for the original file and store it, since transformLineEndings
+  // will modify fileContents buffer.
+  const digestForOriginal = getSingleDigest(fileContents);
+  const transformedFile = transformLineEndings(fileContents, lineEndingFormat);
+  return [digestForOriginal, getSingleDigest(transformedFile)];
+};
+
+const hasLineEndings = (fileContents: Buffer): boolean => {
+  const lineEndingBytes = new Uint8Array([0x0a, 0x0d]);
+  for (let readIndex = 0; readIndex < fileContents.length; readIndex++) {
+    if (lineEndingBytes.includes(fileContents[readIndex])) {
+      return true;
+    }
+  }
+  return false;
+};
+
+const detectLineEndingFormat = (fileContents: Buffer): LineEndingFormat => {
+  for (let readIndex = 0; readIndex < fileContents.length; readIndex++) {
+    // If the first character in the file is LF (0x0a), the file is of type LINUX.
+    if (fileContents[readIndex] === 0x0a && readIndex === 0) {
+      return LineEndingFormat.LINUX;
+    }
+    // If the file has a LF character (0x0a), check if the previous character is 0x0d (CR),
+    // then the newline type is WINDOWS
+    if (
+      fileContents[readIndex] === 0x0a &&
```

Comment:
you can simplify this logic here, not checking if byte is 0x0a in three places

Reply:
Good point. I have refactored this so the multiple checking is not necessary.